File: sdk/samples/FrameworkSamples/NCL/SecureStreams/vb/sslserver/sslserver.vb
'---------------------------------------------------------------------
'  This file is part of the Microsoft .NET Framework SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Imports System.IO
Imports System.Net
Imports System.Net.Sockets
Imports System.Net.Security
Imports System.Security.Authentication
Imports System.Security.Cryptography
Imports System.Security.Cryptography.X509Certificates

Module SslServer

    Sub Main()
        Dim sslStream As SslStream = Nothing
        Try
            ' Set the TcpListener on port 13000.
            Dim port As Integer = 13000
            Dim localAddr As IPAddress = IPAddress.Loopback

            ' TcpListener server = new TcpListener(port);
            Dim server As New TcpListener(localAddr, port)

            ' Start listening for client requests.
            server.Start()

            ' Buffer for reading data
            Dim bytes(256) As Byte
            Dim data As String = Nothing

            ' Enter the listening loop.
            While True
                Console.Write("Waiting for a connection... ")

                ' Perform a blocking call to accept requests.
                ' You could also user server.AcceptSocket() here.
                Dim client As TcpClient = server.AcceptTcpClient()
                Console.WriteLine("Connected!")

                data = Nothing

                ' Get a stream object for reading and writing
                sslStream = New SslStream(client.GetStream())

                ' The exportedcertificate.cer file would contain a 
                ' certificate with public key that came from a certificate
                ' in the local machine store.
                Dim certificate As X509Certificate = _
                       X509Certificate.CreateFromCertFile( _
                       "exportedcertificate.cer")
                sslStream.AuthenticateAsServer(certificate)

                If sslStream.IsAuthenticated Then
                    Console.WriteLine( _
                        "IsAuthenticated: {0}", _
                        sslStream.IsAuthenticated)
                    Console.WriteLine( _
                        "IsMutuallyAuthenticated: {0}", _
                        sslStream.IsMutuallyAuthenticated)
                    Console.WriteLine( _
                        "IsEncrypted: {0}", _
                        sslStream.IsEncrypted)
                    Console.WriteLine( _
                        "IsSigned: {0}", _
                        sslStream.IsSigned)
                    Console.WriteLine( _
                        "IsServer: {0}", _
                        sslStream.IsServer)
                End If

                ' Loop to receive all the data sent by the client.
                Dim bytesRead As Integer = sslStream.Read(bytes, 0, bytes.Length)
                While bytesRead <> 0
                    ' Translate data bytes to a ASCII string.
                    data = System.Text.Encoding.ASCII.GetString( _
                        bytes, 0, bytesRead)
                    Console.WriteLine("Received: {0}", data)

                    ' Process the data sent by the client.
                    data = data.ToUpper( _
                        System.Globalization.CultureInfo.CurrentCulture)

                    Dim msg() As Byte = _
                        System.Text.Encoding.ASCII.GetBytes(data)

                    ' Send back a response.
                    sslStream.Write(msg, 0, msg.Length)
                    Console.WriteLine("Sent: {0}", data)

                    bytesRead = sslStream.Read(bytes, 0, bytes.Length)
                End While
            End While
        Catch ex As CryptographicException
            Console.WriteLine(ex.Message)
            Console.WriteLine( _
                "An error has occurred while using or authenticating " & _
                "this connection.  Check to make sure you've exported " & _
                "your certificate and that it is correctly referenced in " & _
                "CreateFromCertFile above.")
        Catch ex As AuthenticationException
            Console.WriteLine(ex.Message)
        Catch ex As SocketException
            Console.WriteLine(ex.Message)
        Catch ex As IOException
            Console.WriteLine(ex.Message)
        Finally
            If sslStream IsNot Nothing Then
                sslStream.Close()
            End If
        End Try
        Console.WriteLine(Environment.NewLine & "Hit enter to continue...")
        Console.Read()
    End Sub
End Module

File: sdk/samples/FrameworkSamples/NCL/SecureStreams/cs/sslserver/sslserver.cs
//---------------------------------------------------------------------
//  This file is part of the Microsoft .NET Framework SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------

using System;
using System.IO;
using System.Net;
using System.Text;
using System.Net.Sockets;
using System.Net.Security;
using System.Security.Authentication;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;

namespace Microsoft.Samples.SecureStreams
{
    static class SslServer
    {
        public static void Main()
        {
            SslStream sslStream = null;
            try
            {
                // Set the TcpListener on port 13000.
                Int32 port = 13000;
                IPAddress localAddr = IPAddress.Loopback;

                // TcpListener server = new TcpListener(port);
                TcpListener server = new TcpListener(localAddr, port);

                // Start listening for client requests.
                server.Start();

                // Buffer for reading data
                Byte[] bytes = new Byte[256];
                String data = null;

                // Enter the listening loop.
                while (true)
                {
                    Console.Write("Waiting for a connection... ");

                    // Perform a blocking call to accept requests.
                    // You could also user server.AcceptSocket() here.
                    TcpClient client = server.AcceptTcpClient();
                    Console.WriteLine("Connected!");

                    data = null;

                    // Get a stream object for reading and writing
                    sslStream = new SslStream(client.GetStream());

                    // The exportedcertificate.cer file would contain a 
                    // certificate with public key that came from a certificate
                    // in the local machine store.
                    X509Certificate certificate =
                           X509Certificate.CreateFromCertFile(
                           "exportedcertificate.cer");
                    sslStream.AuthenticateAsServer(certificate);

                    if (sslStream.IsAuthenticated)
                    {
                        Console.WriteLine(
                            "IsAuthenticated: {0}",
                            sslStream.IsAuthenticated);
                        Console.WriteLine(
                            "IsMutuallyAuthenticated: {0}",
                            sslStream.IsMutuallyAuthenticated);
                        Console.WriteLine(
                            "IsEncrypted: {0}",
                            sslStream.IsEncrypted);
                        Console.WriteLine(
                            "IsSigned: {0}",
                            sslStream.IsSigned);
                        Console.WriteLine(
                            "IsServer: {0}",
                            sslStream.IsServer);
                    }

                    int i;

                    // Loop to receive all the data sent by the client.
                    while ((i = sslStream.Read(bytes, 0, bytes.Length)) != 0)
                    {
						// Translate data bytes to a string.
						// The encoding used is application specific.
						data =
							System.Text.Encoding.ASCII.GetString(bytes, 0, i);
                        Console.WriteLine("Received: {0}", data);

                        // Process the data sent by the client.
                        data = data.ToUpper(
                            System.Globalization.CultureInfo.CurrentCulture);

                        byte[] msg =
                            System.Text.Encoding.ASCII.GetBytes(data);

                        // Send back a response.
                        sslStream.Write(msg, 0, msg.Length);
                        Console.WriteLine("Sent: {0}", data);
                    }
                }
            }
            catch (CryptographicException ex)
            {
                Console.WriteLine(ex.Message);
                Console.WriteLine(
                    "An error has occurred while using or authenticating " +
                    "this connection.  Check to make sure you've exported " +
                    "your certificate and that it is correctly referenced " +
                    "in CreateFromCertFile above.");
            }
            catch (AuthenticationException ex)
            {
                Console.WriteLine(ex.Message);
            }
            catch (SocketException ex)
            {
                Console.WriteLine(ex.Message);
            }
            catch (IOException ex)
            {
                Console.WriteLine(ex.Message);
            }
            finally
            {
                if (sslStream != null)
                    sslStream.Close();
            }

            Console.WriteLine("\nHit enter to continue...");
            Console.Read();
        }
    }

}
